### PR TITLE
Run multiple runs in paralell

### PR DESCRIPTION
**Pull request by matalbec**
Includes data store refactor and a very initial implementation of multiprocessing.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,7 +7,6 @@
 *.log
 src/__pycache__*
 .idea*
-src/plots/*
```
Comment (MacStan) [inline .gitignore:10]: good one

File: src/data_storage.py [hunk 2]
```diff
@@ -4,75 +4,65 @@
 import sys
 from plotter import Plotter
 
+class RunStoreHandle:
```
Comment (MacStan) [inline src/data_storage.py:7]: It's Handler not Handle, isn't it?

File: src/image_generator.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import sys
+from plotter import Plotter
+
+class ImageGenerator():
+    def __init__(self):
```
Comment (MacStan) [inline src/image_generator.py]: It seems like **new_path** should be a field, not argument of methods.

File: src/runner.py [hunk 4]
```diff
@@ -1,16 +1,29 @@
...
-            print("FAILURE DURING RUN")
-            traceback.print_exc()
+    for seed in args.seeds:
+        for size_adj in args.size_adjs:
+            processInputArgs.append([
+                RunParams(seed, args.steps, size_adj, data, args.freq, args.tag),
+                dataStore.open_run_store(runIndex)])
+            runIndex += 1
+
+    pool = Pool()
```
Comment (MacStan) [inline src/runner.py:25]: We would prefer smarter solution than just running everything in parallel, wouldn't we? :P Not an issue though

File: src/data_storage.py [hunk 5]
```diff
@@ -4,75 +4,65 @@
...
+        self.runStorePath = runDesc["runStorePath"]
+        self.configFilePath = os.path.join(self.runStorePath, "net-config")
+        self.dataFilePath = os.path.join(self.runStorePath, "data-points")
+        if not os.path.exists(self.runStorePath):
+            os.makedirs(self.runStorePath)
+        self.runData = self.load_run_data()
 
-class DataStorage:
-    """Storage creator.
-    Should be used once per given path, so for each new 'run' new DataStorage should be created."""
```
Comment (MacStan) [inline src/data_storage.py:10]: Comment was still valid, though?

File: src/data_storage.py [hunk 6]
```diff
@@ -4,75 +4,65 @@
...
-            configFile.write(json.dumps(net_conf))
+    def commit_data(self):
+        with open(self.dataFilePath, 'w') as dataFile:
+            dataFile.write(json.dumps(self.runData))
 
-    def save_data_point(self, net_state, point_index):
-        with open(os.path.join(self.storage_path, f"data-point{point_index}"), 'x') \
-                as dataFile:
-            dataFile.write(json.dumps(net_state))
+    def read_data_point(self, dataPointIndex):
```
Comment (MacStan) [inline src/data_storage.py]: Read implies reading from disc, what do you say about "get"?

File: src/data_storage.py [hunk 7]
```diff
@@ -4,75 +4,65 @@
...
+                return json.loads(dataFile.read())
+        else:
+            return []
 
-    def get_data_point(self, seed, point_index):
-        with open(os.path.join(self.runStoragePath, f"seed{seed}/data-point{point_index}"),
-                  'r') as dataFile:
-            return json.loads(dataFile.read())
+    def get_run_length(self):
+        return len(self.runData)
```
Comment (MacStan) [inline src/data_storage.py]: Maybe override __len__ method?

File: src/image_generator.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import sys
```
Comment (MacStan) [inline src/image_generator.py:1]: Nice addition :)

File: src/runner.py [hunk 9]
```diff
@@ -1,16 +1,29 @@
...
+    pool.map(run_wrapper, processInputArgs)
```
Comment (MacStan) [inline src/runner.py:26]: Maybe print some info about processes? Amount, currently running params?

**Review by MacStan — CHANGES_REQUESTED** [review]
Please review ImageGenerator and len() in RunStoreHandler

File: src/data_storage.py (hunk 2 above)
Comment (matalbec) [reply src/data_storage.py:7]: It's a handle. Similarly to the terminology used by files, not as robust though

File: src/data_storage.py (hunk 5 above)
Comment (matalbec) [reply src/data_storage.py:10]: Nope, completely new model is used.

File: src/data_storage.py (hunk 2 above)
Comment (MacStan) [reply src/data_storage.py:7]: Ok, didn't read code thorougly enough

File: src/data_storage.py (hunk 5 above)
Comment (MacStan) [reply src/data_storage.py:10]: OK

File: src/data_storage.py (hunk 7 above)
Comment (MacStan) [reply src/data_storage.py]: I meant, add __ len __ to Handle

**Review by MacStan — CHANGES_REQUESTED** [review]

File: src/data_storage.py (hunk 7 above)
Comment (matalbec) [reply src/data_storage.py]: sure

File: src/runner.py (hunk 9 above)
Comment (matalbec) [reply src/runner.py:26]: Those are printed from inside the process. A little messy but it will get better once logging is fixed.

**Review by MacStan — APPROVED** [review]
Very nice!

**Pull request merged**